backend/src/routes/index.js

```diff
@@ -44,15 +47,22 @@ router.get('/dashboard*', [requireAuth], (req, res) => {
   return res.render('admin');
 });
 
-router.post('/login', login);
+router.post('/login', passport.authenticate('local', { successRedirect: '/dashboard', failureRedirect: '/login' }));
 router.get('/logout', (req, res) =>
   req.session.destroy(() => res.redirect('/login'))
 );
 
-router.get('/admins', [requireAuth, superAdminOnly], adminController.list);
-router.post('/admins', [requireAuth, superAdminOnly], adminController.createSuperAdmin);
-router.put('/admins/:adminId', [requireAuth, superAdminOnly], adminController.updateSuperAdmin);
-router.delete('/admins/:adminId', [requireAuth, superAdminOnly], adminController.deleteSuperAdmin);
+// These matchers/middlewares greatly reduce the need for entity existence checks in controllers
+router.all('*/branches/:branchId*', resourceValidators.checkBranchPresence);
+router.all('*/admins/:adminId*', resourceValidators.checkAdminPresence);
+router.all('*/branches/:branchId/admins/:adminId*', resourceValidators.checkBranchAdminPresence);
```

Comment:
The stars are wildcards, so that these routes match requests to any endpoint containing the middle bit. So any request that contains `/branches/123-456` will go through the middleware that checks that branch 123-456 exists. It means that you don't need to manually specify this middleware for `/branches/:branchId/admins` and `/branches/:branchId/members/:memberId` etc etc etc.

Previously these kind of validations were copy-pasted across all the various controllers. Now they can be done in once place 👍